Expose inner `GcRefCell` as `CollectRefCell` and also add `CollectCell`

kyren opened the pull request:
These types are generally useful, even if they require unsafe to be useful.

It's nice to not have to manually make an unsafe `Collect` impl, rather than only having to remember to trigger a write barrier.

kyren committed:
Expose inner `GcRefCell` as `CollectRefCell` and also add `CollectCell`

These types are generally useful, even if they require unsafe to be
useful.

It's nice to not have to manually make an unsafe `Collect` impl, rather
than only having to remember to trigger a write barrier.

kyren commented:
If anyone has a better name than `CollectRefCell` and `CollectCell` I'm all ears.

kyren committed:
remove exclamation in one comment but not the other that made things more confusing

kyren committed:
Completely get rid of GcCell and GcWeakCell

They are not necessary anymore, because it is very straightforward
to just provide safe methods on the `Gc<CollectCell>` and
`Gc<CollectRefCell>` types directly.

This results in slightly more verbose code, but it is vastly simpler,
and if we can have projected `Gc` pointer types, this or something
similar to this should enable fairly ergonomic fine grained mutability.

moulins commented:
Instead of having a separate method for each unsafe operation on each cell type, I'd propose just providing a single method:
```rs
unsafe fn as_(ref_)cell(&self) -> &(Ref)Cell<T>;
```

This is slightly more verbose to use, but IMHO, it has two advantages:
- The user can call any method on the `Cell/RefCell`, not just those we explicitely support;
- Using a specific method name clarifies what "kind" of unsafety is happening.

---------

> If anyone has better names than `CollectRefCell` and `CollectCell` I'm all ears.

Some suggestions:
- `Lock/RefLock`: this is what I'm using in my [`gc-extra` experiment](https://github.com/moulins/ruffle/tree/gc-extra/gc_extra/src/lock);
- `Jail/RefJail`: cute variation on `Cell`;
- `FrozenCell/FrozenRefCell`: a little verbose, but maybe more explicit?

moulins commented on src/gc-arena/src/cell.rs:
```diff
@@ -0,0 +1,152 @@
+    }
+}
+
+unsafe impl<'gc, T: Collect + Copy + 'gc> Collect for CollectCell<T> {
+    fn trace(&self, cc: CollectionContext) {
+        self.cell.get().trace(cc);
+    }
+}
```
This is a very degenerate case, but this is incorrect if `T::trace` modifies the inner `T`; the potentially modified value should be written back to the `Cell`:
```rs
fn trace(&self, cc: CollectionContext) {
    // Use a guard that writes back the value in case the 'trace' call mutates it.
    let cell = &self.cell;
    let guard = scopeguard::guard(cell.get(), |copied| cell.set(copied));
    T::trace(&guard, cc);
}
```

kyren committed:
Rename Collect[Ref]Cell to [Ref]Lock, and just provide direct access to the inner cells

kyren committed:
Be correct in the face of somewhat pathalogical trace behavior in a Lock

kyren committed:
Remove the code that handles `Lock` pathology, it is impossible to observe

Right now, there are not types that have interior mutability and
are Copy, so you actually can't possibly mutate a contained type
in a `Lock`. Instead, I left a long note about the (currently only
theoretical) issue.
